{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "SERVER = \"127.0.0.1\"\n",
    "PORT = 1234\n",
    "USERNAME = \"username\"\n",
    "PASSWORD = \"password\"\n",
    "SSL = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "EXPORT_DIR = \"/mnt/d/Documents/tmp\"\n",
    "EXPORT_FORMAT = \"CVAT for video 1.1\"\n",
    "EXPORT_INCLUDE_IMAGES = False\n",
    "\n",
    "if not os.path.exists(EXPORT_DIR):\n",
    "    os.makedirs(EXPORT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/112 [00:00<?, ?it/s]ERROR:root:Failed to export task train_hiber2_43.mp4, id 60: Traceback (most recent call last):\n",
      "  File \"/tmp/ipykernel_317/3971423106.py\", line 31, in <module>\n",
      "    client.tasks.retrieve(obj_id=task.id).export_dataset(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/proxies/tasks.py\", line 234, in export_dataset\n",
      "    Downloader(self._client).prepare_and_download_file_from_endpoint(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/downloading.py\", line 89, in prepare_and_download_file_from_endpoint\n",
      "    client.wait_for_completion(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/client.py\", line 187, in wait_for_completion\n",
      "    response = self.api_client.rest_client.request(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/api_client/rest.py\", line 209, in request\n",
      "    raise ServiceException(http_resp=r)\n",
      "cvat_sdk.api_client.exceptions.ServiceException: Status Code: 500\n",
      "Reason: Internal Server Error\n",
      "HTTP response headers: HTTPHeaderDict({'Server': 'nginx/1.23.3', 'Date': 'Sun, 19 Mar 2023 05:48:00 GMT', 'Content-Type': 'application/vnd.cvat+json', 'Content-Length': '23', 'Connection': 'keep-alive', 'Allow': 'GET, POST, PUT, PATCH, DELETE, HEAD, OPTIONS', 'Referrer-Policy': 'same-origin', 'Vary': 'Accept,Origin', 'X-Content-Type-Options': 'nosniff', 'X-Frame-Options': 'DENY'})\n",
      "HTTP response body: b'\"KeyError: \\'outside\\'\\\\n\"'\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/tmp/ipykernel_317/3971423106.py\", line 31, in <module>\n",
      "    client.tasks.retrieve(obj_id=task.id).export_dataset(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/proxies/tasks.py\", line 234, in export_dataset\n",
      "    Downloader(self._client).prepare_and_download_file_from_endpoint(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/downloading.py\", line 89, in prepare_and_download_file_from_endpoint\n",
      "    client.wait_for_completion(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/client.py\", line 187, in wait_for_completion\n",
      "    response = self.api_client.rest_client.request(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/api_client/rest.py\", line 209, in request\n",
      "    raise ServiceException(http_resp=r)\n",
      "cvat_sdk.api_client.exceptions.ServiceException: Status Code: 500\n",
      "Reason: Internal Server Error\n",
      "HTTP response headers: HTTPHeaderDict({'Server': 'nginx/1.23.3', 'Date': 'Sun, 19 Mar 2023 05:48:00 GMT', 'Content-Type': 'application/vnd.cvat+json', 'Content-Length': '23', 'Connection': 'keep-alive', 'Allow': 'GET, POST, PUT, PATCH, DELETE, HEAD, OPTIONS', 'Referrer-Policy': 'same-origin', 'Vary': 'Accept,Origin', 'X-Content-Type-Options': 'nosniff', 'X-Frame-Options': 'DENY'})\n",
      "HTTP response body: b'\"KeyError: \\'outside\\'\\\\n\"'\n",
      "\n",
      " 66%|██████▌   | 74/112 [00:04<00:02, 17.87it/s]ERROR:root:Failed to export task train_hiber2_42.mp4, id 59: Traceback (most recent call last):\n",
      "  File \"/tmp/ipykernel_317/3971423106.py\", line 31, in <module>\n",
      "    client.tasks.retrieve(obj_id=task.id).export_dataset(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/proxies/tasks.py\", line 234, in export_dataset\n",
      "    Downloader(self._client).prepare_and_download_file_from_endpoint(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/downloading.py\", line 89, in prepare_and_download_file_from_endpoint\n",
      "    client.wait_for_completion(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/client.py\", line 187, in wait_for_completion\n",
      "    response = self.api_client.rest_client.request(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/api_client/rest.py\", line 209, in request\n",
      "    raise ServiceException(http_resp=r)\n",
      "cvat_sdk.api_client.exceptions.ServiceException: Status Code: 500\n",
      "Reason: Internal Server Error\n",
      "HTTP response headers: HTTPHeaderDict({'Server': 'nginx/1.23.3', 'Date': 'Sun, 19 Mar 2023 05:48:05 GMT', 'Content-Type': 'application/vnd.cvat+json', 'Content-Length': '23', 'Connection': 'keep-alive', 'Allow': 'GET, POST, PUT, PATCH, DELETE, HEAD, OPTIONS', 'Referrer-Policy': 'same-origin', 'Vary': 'Accept,Origin', 'X-Content-Type-Options': 'nosniff', 'X-Frame-Options': 'DENY'})\n",
      "HTTP response body: b'\"KeyError: \\'outside\\'\\\\n\"'\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/tmp/ipykernel_317/3971423106.py\", line 31, in <module>\n",
      "    client.tasks.retrieve(obj_id=task.id).export_dataset(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/proxies/tasks.py\", line 234, in export_dataset\n",
      "    Downloader(self._client).prepare_and_download_file_from_endpoint(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/downloading.py\", line 89, in prepare_and_download_file_from_endpoint\n",
      "    client.wait_for_completion(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/client.py\", line 187, in wait_for_completion\n",
      "    response = self.api_client.rest_client.request(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/api_client/rest.py\", line 209, in request\n",
      "    raise ServiceException(http_resp=r)\n",
      "cvat_sdk.api_client.exceptions.ServiceException: Status Code: 500\n",
      "Reason: Internal Server Error\n",
      "HTTP response headers: HTTPHeaderDict({'Server': 'nginx/1.23.3', 'Date': 'Sun, 19 Mar 2023 05:48:05 GMT', 'Content-Type': 'application/vnd.cvat+json', 'Content-Length': '23', 'Connection': 'keep-alive', 'Allow': 'GET, POST, PUT, PATCH, DELETE, HEAD, OPTIONS', 'Referrer-Policy': 'same-origin', 'Vary': 'Accept,Origin', 'X-Content-Type-Options': 'nosniff', 'X-Frame-Options': 'DENY'})\n",
      "HTTP response body: b'\"KeyError: \\'outside\\'\\\\n\"'\n",
      "\n",
      "ERROR:root:Failed to export task train_hiber2_41.mp4, id 58: Traceback (most recent call last):\n",
      "  File \"/tmp/ipykernel_317/3971423106.py\", line 31, in <module>\n",
      "    client.tasks.retrieve(obj_id=task.id).export_dataset(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/proxies/tasks.py\", line 234, in export_dataset\n",
      "    Downloader(self._client).prepare_and_download_file_from_endpoint(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/downloading.py\", line 89, in prepare_and_download_file_from_endpoint\n",
      "    client.wait_for_completion(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/client.py\", line 187, in wait_for_completion\n",
      "    response = self.api_client.rest_client.request(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/api_client/rest.py\", line 209, in request\n",
      "    raise ServiceException(http_resp=r)\n",
      "cvat_sdk.api_client.exceptions.ServiceException: Status Code: 500\n",
      "Reason: Internal Server Error\n",
      "HTTP response headers: HTTPHeaderDict({'Server': 'nginx/1.23.3', 'Date': 'Sun, 19 Mar 2023 05:48:09 GMT', 'Content-Type': 'application/vnd.cvat+json', 'Content-Length': '23', 'Connection': 'keep-alive', 'Allow': 'GET, POST, PUT, PATCH, DELETE, HEAD, OPTIONS', 'Referrer-Policy': 'same-origin', 'Vary': 'Accept,Origin', 'X-Content-Type-Options': 'nosniff', 'X-Frame-Options': 'DENY'})\n",
      "HTTP response body: b'\"KeyError: \\'outside\\'\\\\n\"'\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/tmp/ipykernel_317/3971423106.py\", line 31, in <module>\n",
      "    client.tasks.retrieve(obj_id=task.id).export_dataset(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/proxies/tasks.py\", line 234, in export_dataset\n",
      "    Downloader(self._client).prepare_and_download_file_from_endpoint(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/downloading.py\", line 89, in prepare_and_download_file_from_endpoint\n",
      "    client.wait_for_completion(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/client.py\", line 187, in wait_for_completion\n",
      "    response = self.api_client.rest_client.request(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/api_client/rest.py\", line 209, in request\n",
      "    raise ServiceException(http_resp=r)\n",
      "cvat_sdk.api_client.exceptions.ServiceException: Status Code: 500\n",
      "Reason: Internal Server Error\n",
      "HTTP response headers: HTTPHeaderDict({'Server': 'nginx/1.23.3', 'Date': 'Sun, 19 Mar 2023 05:48:09 GMT', 'Content-Type': 'application/vnd.cvat+json', 'Content-Length': '23', 'Connection': 'keep-alive', 'Allow': 'GET, POST, PUT, PATCH, DELETE, HEAD, OPTIONS', 'Referrer-Policy': 'same-origin', 'Vary': 'Accept,Origin', 'X-Content-Type-Options': 'nosniff', 'X-Frame-Options': 'DENY'})\n",
      "HTTP response body: b'\"KeyError: \\'outside\\'\\\\n\"'\n",
      "\n",
      " 69%|██████▉   | 77/112 [00:16<00:10,  3.35it/s]ERROR:root:Failed to export task train_hiber2_39.mp4, id 56: Traceback (most recent call last):\n",
      "  File \"/tmp/ipykernel_317/3971423106.py\", line 31, in <module>\n",
      "    client.tasks.retrieve(obj_id=task.id).export_dataset(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/proxies/tasks.py\", line 234, in export_dataset\n",
      "    Downloader(self._client).prepare_and_download_file_from_endpoint(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/downloading.py\", line 89, in prepare_and_download_file_from_endpoint\n",
      "    client.wait_for_completion(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/client.py\", line 185, in wait_for_completion\n",
      "    sleep(status_check_period)\n",
      "KeyboardInterrupt\n",
      "Traceback (most recent call last):\n",
      "  File \"/tmp/ipykernel_317/3971423106.py\", line 31, in <module>\n",
      "    client.tasks.retrieve(obj_id=task.id).export_dataset(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/proxies/tasks.py\", line 234, in export_dataset\n",
      "    Downloader(self._client).prepare_and_download_file_from_endpoint(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/downloading.py\", line 89, in prepare_and_download_file_from_endpoint\n",
      "    client.wait_for_completion(\n",
      "  File \"/home/thinszx/miniconda3/envs/misc/lib/python3.10/site-packages/cvat_sdk/core/client.py\", line 185, in wait_for_completion\n",
      "    sleep(status_check_period)\n",
      "KeyboardInterrupt\n",
      " 70%|██████▉   | 78/112 [00:22<00:17,  1.94it/s]"
     ]
    }
   ],
   "source": [
    "from cvat_sdk import make_client\n",
    "from cvat_sdk.core.helpers import TqdmProgressReporter # https://github.com/opencv/cvat/blob/develop/cvat-cli/src/cvat_cli/cli.py#L12\n",
    "import tqdm\n",
    "\n",
    "if SSL:\n",
    "    HOST = f\"https://{SERVER}\"\n",
    "else:\n",
    "    HOST = f\"http://{SERVER}\"\n",
    "\n",
    "with make_client(host=HOST, port=F\"{PORT}\", credentials=(f\"{USERNAME}\", f\"{PASSWORD}\")) as client:\n",
    "    tasks = client.tasks.list()\n",
    "    # jobs = client.jobs.list()\n",
    "    for task in tqdm.tqdm(tasks):\n",
    "        # pbar = TqdmProgressReporter(\n",
    "        #         tqdm.tqdm_notebook(unit_scale=True, unit=\"B\", unit_divisor=1024, desc=task.name)\n",
    "        #     )\n",
    "        if task.id > 60:\n",
    "            continue\n",
    "        try:\n",
    "            pbar = None\n",
    "            client.tasks.retrieve(obj_id=task.id).export_dataset(\n",
    "                    format_name=EXPORT_FORMAT,\n",
    "                    filename=os.path.join(EXPORT_DIR, f\"{task.name}.zip\"),\n",
    "                    pbar=pbar,\n",
    "                    status_check_period=2,\n",
    "                    include_images=EXPORT_INCLUDE_IMAGES,\n",
    "                )\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "misc",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
